flow/flow.go

```diff
@@ -0,0 +1,291 @@
+	// other unique value).
+	//
+	// If you fail to compute the proper value, then authentication processes which have id_token_hint set might fail.
+	ForceSubjectIdentifier string `json:"force_subject_identifier" db:"forced_subject_identifier"`
+
+	// Context is an optional object which can hold arbitrary data. The data will be made available when fetching the
+	// consent request under the "context" field. This is useful in scenarios where login and consent endpoints share
+	// data.
+	Context sqlxx.JSONRawMessage `json:"context" db:"context"`
+
+	// If set to true means that the request was already handled. This
+	// can happen on form double-submit or other errors. If this is set
+	// we recommend redirecting the user to `request_url` to re-initiate
+	// the flow.
+	WasHandled bool `json:"-" db:"was_used"`
+
+	Error                *consent.RequestDeniedError `json:"-" db:"error"`
+	LoginAuthenticatedAt sqlxx.NullTime              `json:"-" db:"login_authenticated_at"`
+}
+
+func NewFlow(r *consent.LoginRequest) *Flow {
+	return &Flow{
+		ID:                     r.ID,
+		RequestedScope:         r.RequestedScope,
+		RequestedAudience:      r.RequestedAudience,
+		Skip:                   r.Skip,
+		Subject:                r.Subject,
+		OpenIDConnectContext:   r.OpenIDConnectContext,
+		Client:                 r.Client,
+		ClientID:               r.ClientID,
+		RequestURL:             r.RequestURL,
+		SessionID:              r.SessionID,
+		WasHandled:             r.WasHandled,
+		ForceSubjectIdentifier: r.ForceSubjectIdentifier,
+		Verifier:               r.Verifier,
+		CSRF:                   r.CSRF,
+		LoginAuthenticatedAt:   r.AuthenticatedAt,
+		RequestedAt:            r.RequestedAt,
+		State:                  FlowStateLoginInitialized,
+	}
+}
+
+func (f *Flow) HandleLoginRequest(h *consent.HandledLoginRequest) error {
+	if f.State != FlowStateLoginInitialized {
+		return errors.New(fmt.Sprintf("Invalid state for HandleLoginRequest. Expected %d, got %d", FlowStateLoginUnused, f.State))
+	}
+	if f.ID != h.ID {
+		return errors.New(fmt.Sprintf("Flow ID %s does not match HandledLoginRequest ID %s", f.ID, h.ID))
+	}
+	// Each of the below checks causes some tests to fail. TODO determine whether these conditions require special care
+	//if f.Subject != h.Subject {
+	//	return errors.New(fmt.Sprintf("Flow Subject %s does not match the HandledLoginRequest Subject %s", f.Subject, h.Subject))
+	//}
+	//if f.ForceSubjectIdentifier != h.ForceSubjectIdentifier {
+	//	return errors.New(fmt.Sprintf("Flow ForceSubjectIdentifier %s does not match the HandledLoginRequest ForceSubjectIdentifier  %s", f.ForceSubjectIdentifier, h.ForceSubjectIdentifier))
+	//}
+	//if h.Error != nil {
+	//	return errors.New("Refusing to process HandledLoginRequest " + h.ID + " with error")
+	//}
+
+	f.State = FlowStateLoginUnused
+	f.ID = h.ID
+	f.Subject = h.Subject
```

Comment:
Do we need Subject and ForceSubjectIdentifier in the LoginRequest if we overwrite it here?